### PR TITLE
Thread pool implementation

### PR DESCRIPTION
**Pull request by st-pasha**
This PR adds an internal thread pool + supporting classes, which will eventually be used to replace OMP parallelism.

Currently supported functionality:
- increase/reduce number of threads in the pool;
- `dt::run_once_per_thread()` function can be used for plain parallelization (similar to plain `#pragma omp parallel`);
- generic scheduler mechanism, which can be extended to implement other types of parallelism, such as parallel for or parallel ordered for;
- python function `dt.internal.get_thread_ids()` return a list of strings which are implementation-dependent thread ids (can be used for debugging, maybe);
- added macro `CACHELINE_SIZE` and struct `cache_aligned<T>`.

WIP for #1736

### CODE REVIEW

File: c/parallel/thread_scheduler.cc [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  return &tsleep[index];
+}
+
+void thread_sleep_scheduler::wait_until_finish() {
+  // This is the only non-joinable task. Don't do anything here.
+}
+
+void thread_sleep_scheduler::awaken(thread_scheduler* next) {
+  size_t i = index;
+  size_t j = (i + 1) & 1;
```
Comment (oleksiyskononenko) [inline c/parallel/thread_scheduler.cc]: What is this `j` index?

File: c/parallel/thread_scheduler.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ * The master thread that called `awaken(job)` will then call `job.join()`,
+ * and it is the responsibility of thread_scheduler `job` to wait until all
+ * threads have finished execution and were put back to sleep. Thus, the master
+ * thread ensures that all threads are sleeping again before the next call to
+ * `awaken`.
+ */
+class thread_sleep_scheduler : public thread_scheduler {
+  private:
+    thread_sleep_task tsleep[2];
+    size_t index = 0;  // switches between 0 or 1
```
Comment (oleksiyskononenko) [inline c/parallel/thread_scheduler.h]: Could you explain what this `index` is? Btw, if it only takes 0/1 values, may be we make it `bool`?

File: c/datatablemodule.cc [hunk 3]
```diff
@@ -142,6 +145,28 @@ static py::oobj has_omp_support(const py::PKArgs&) {
 }
 
 
+static py::PKArgs args_get_thread_ids(
+    0, 0, 0, false, false, {}, "get_thread_ids",
+R"(Return internal ids of all threads used internally by datatable)");
```
[the remark this answers is not in the document]
Comment (st-pasha) [reply c/datatablemodule.cc]: fixed

File: c/parallel/thread_scheduler.cc (hunk 1 above)
Comment (st-pasha) [reply c/parallel/thread_scheduler.cc]: see my new commit

**Pull request merged**